### ISSUE TITLE
Gather backward is incorrect with repeated indices

### ISSUE THREAD

**Issue by jcjohnson** [pytorch/pytorch#1631]
The gather function gives incorrect gradients on both CPU and GPU when using repeated indices; no warnings or errors are raised, and the documentation doesn't say anything about this. I've seen this discussed a bit on slack, but couldn't find any issues for it on GitHub.

Here's a small test case (using PyTorch 0.1.12_2 on Ubuntu 16.04):

```python
import torch
from torch.autograd import Variable

def gather_test(dtype):
  x = torch.FloatTensor([0.1])
  i = torch.LongTensor([0, 0])
  x = Variable(x.type(dtype), requires_grad=True)
  i = Variable(i.type(dtype).long())
  y = x.gather(0, i)
  dy = torch.FloatTensor([0.2, 0.3]).type(dtype)
  y.backward(dy)
  print('dtype: ', dtype)
  print('x: ', x.data.cpu().numpy())
  print('i: ', i.data.cpu().numpy())
  print('y: ', y.data.cpu().numpy())
  print('dy: ', dy.cpu().numpy())
  print('dx: ', x.grad.data.cpu().numpy())

gather_test(torch.FloatTensor)
gather_test(torch.cuda.FloatTensor)
```

This gives the output:

```
dtype:  <class 'torch.FloatTensor'>
x:  [ 0.1]
i:  [0 0]
y:  [ 0.1  0.1]
dy:  [ 0.2         0.30000001]
dx:  [ 0.30000001]
dtype:  <class 'torch.cuda.FloatTensor'>
x:  [ 0.1]
i:  [0 0]
y:  [ 0.1  0.1]
dy:  [ 0.2         0.30000001]
dx:  [ 0.30000001]
```

For this example I expect dx to be [0.5] since the first element of x is duplicated to both elements of y; however instead of summing the elements of dy, the gather backward pass simply chooses one of the elements of dy corresponding to the element of x.

In this particular example we could use `index_select` instead of `gather`, but there are a lot of situations where `index_select` cannot substitute for `gather`; it would thus be extremely useful if `gather` could support repeated indices.

**Comment by jcjohnson** [comment]
It seems to be a problem with scatter. Gather backward is implemented using scatter:

https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/tensor.py#L511

CPU scatter just overwrites the destination:

https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L465
https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L488

Same story for GPU scatter:

https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorScatterGather.cu#L126
https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorScatterGather.cu#L151

I think we could add an optional flag to scatter to have it accumulate into the destination tensor rather then overwrite, then use the accumulation version of scatter in the gather backward pass.

Unfortunately I don't think I'm familiar enough with the TH / THC guts to fix this myself.

**Comment by fmassa** [comment]
@jcjohnson you are right, and Adam just sent a PR to fix this issue

**Comment by jcjohnson** [comment]
Hah, looks like that's exactly what @adamlerer did in https://github.com/pytorch/pytorch/pull/1632
